### ISSUE TITLE
Difficulty reproducing initialization depth

### ISSUE THREAD

**Issue by m43** [ImNotPrepared/MonoFusion#3]
I’ve tried reproducing your initialization depth, but generally find it doesn’t give reasonable reconstructions. May I see how your initialization depth looks like on the “basketball” Panoptic sequence? Below is what I get with different methods

<table>
  <tr>
    <th>Dust3r</th>
    <th>Moge2 (metric)</th>
    <th>Dust3r + Moge2</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/bfaab926-e571-41a4-9b18-49aa21bd46e4" width="333"/></td>
    <td><img src="https://github.com/user-attachments/assets/86d92467-5801-40ad-8b88-e4c85d1ccb58" width="333"/></td>
    <td><img src="https://github.com/user-attachments/assets/561f8a43-03bc-49a0-a121-b30fbaf59019" width="333"/></td>
  </tr>
</table>

Below are the relevant parts of my depth estimation scripts for your reference, note that I have used a simpler foreground-background heuristic that works on Panoptic instead of Sam2:
```py
def _ensure_moge2_cache_and_load(rgbs, seq_name, dataset_root, moge2_cache_subdir, skip_if_cached=True):
    """
    Raw MoGe-2 depth (metric) with per-sequence caching.
    Returns (depths, confs) shaped [V,T,1,H,W] on CPU.
    """
    V, T, _, H, W = rgbs.shape
    cache_root = os.path.join(dataset_root, moge2_cache_subdir, seq_name)
    os.makedirs(cache_root, exist_ok=True)
    depths_path = os.path.join(cache_root, "moge2_depths.npy")
    confs_path = os.path.join(cache_root, "moge2_confs.npy")

    if skip_if_cached and os.path.isfile(depths_path) and os.path.isfile(confs_path):
        d = torch.from_numpy(np.load(depths_path)).float()  # [V,T,H,W]
        c = torch.from_numpy(np.load(confs_path)).float()  # [V,T,H,W]
        return d.unsqueeze(2), c.unsqueeze(2)

    d = _moge_depths(seq_name, rgbs, cache_root)  # [V,T,H,W], CPU float

    # Simple constant confidence for MoGe-2
    c = torch.full_like(d, 100.0)

    np.save(depths_path, d.numpy())
    np.save(confs_path, c.numpy())
    return d.unsqueeze(2), c.unsqueeze(2)


def _ensure_monofusion_cache_and_load(rgbs, seq_name, dataset_root, monofusion_cache_subdir, skip_if_cached=True):
    """
    MONOFUSION:
      - Background mask: patch-change detector over temporal window (static -> BG)
      - DUSt3R depth: load per frame/view; build static background depth by BG-temporal-average.
      - MoGe-2 monocular depth per frame/view; align to background by affine (a,b).
      - Merge BG (DUSt3R static) with FG (aligned MoGe).
      - Cache final depths & confs.
    """
    V, T, _, H, W = rgbs.shape

    cache_root = os.path.join(dataset_root, monofusion_cache_subdir, seq_name)
    os.makedirs(cache_root, exist_ok=True)
    final_depths_path = os.path.join(cache_root, "final_depths.npy")
    final_confs_path = os.path.join(cache_root, "final_confs.npy")

    if skip_if_cached and os.path.isfile(final_depths_path) and os.path.isfile(final_confs_path):
        fd = torch.from_numpy(np.load(final_depths_path))  # [V,T,H,W]
        fc = torch.from_numpy(np.load(final_confs_path))  # [V,T,H,W]
        return fd.unsqueeze(2), fc.unsqueeze(2)

    # ---- DUSt3R depths per frame/view ----
    depth_root = os.path.join(dataset_root, f"duster_depths__{seq_name}")
    if not os.path.exists(os.path.join(depth_root, f"3d_model__{T - 1:05d}__scene.npz")):
        if "../duster" not in sys.path:
            sys.path.insert(0, "../duster")
        from scripts.egoexo4d_preprocessing import main_estimate_duster_depth
        pkl_path = os.path.join(dataset_root, f"{seq_name}.pkl")

        # Re-enable autograd locally (overrides any surrounding no_grad/inference_mode)
        with ExitStack() as stack:
            stack.enter_context(torch.inference_mode(False))
            stack.enter_context(torch.enable_grad())
            main_estimate_duster_depth(pkl_path, depth_root, skip_if_cached)

    duster_depths = []
    for t in range(T):
        scene_path = os.path.join(depth_root, f"3d_model__{t:05d}__scene.npz")
        scene = np.load(scene_path)
        d = torch.from_numpy(scene["depths"])  # [V, H', W']
        d = interpolate(d[:, None], size=(H, W), mode="nearest")[:, 0]  # [V, H, W]
        duster_depths.append(d)
    duster_depths = torch.stack(duster_depths, dim=1)  # [V, T, H, W]

    # ---- Background mask (patch-change) ----
    compute_device = "cuda" if torch.cuda.is_available() else "cpu"
    bg_mask = _static_bg_mask_from_window(rgbs.to(compute_device)).cpu()  # [V,T,H,W] bool

    # ---- Static background depth per camera via temporal average on BG pixels ----
    V, T, _, _ = duster_depths.shape
    D_bg = torch.zeros((V, H, W), dtype=torch.float32)
    for v in range(V):
        valid = bg_mask[v]  # [T,H,W]
        num = (duster_depths[v] * valid).sum(dim=0)
        den = valid.sum(dim=0).clamp_min(1)
        D_bg[v] = num / den

    # ---- MoGe-2 monocular depths per frame/view ----
    moge_depths = _moge_depths(seq_name, rgbs, cache_root)  # [V,T,H,W]

    # ---- Align MoGe to background (solve a,b on BG pixels) ----
    compute_device = "cuda" if torch.cuda.is_available() else "cpu"
    moge_depths = moge_depths.to(compute_device, dtype=torch.float32)  # [V,T,H,W]
    D_bg_exp = D_bg[:, None].expand_as(moge_depths).to(compute_device)  # [V,T,H,W]
    bg_mask = bg_mask.to(compute_device)  # [V,T,H,W]

    # Valid BG pixels
    valid = bg_mask & torch.isfinite(moge_depths) & (moge_depths > 0) \
            & torch.isfinite(D_bg_exp) & (D_bg_exp > 0)

    # Flatten over pixels
    X = moge_depths.view(V, T, -1)  # [V,T,HW]
    Y = D_bg_exp.view(V, T, -1)  # [V,T,HW]
    M = valid.view(V, T, -1).float()  # [V,T,HW]

    # Count valid pixels
    n = M.sum(dim=-1)  # [V,T]
    min_bg = 200
    if (n < min_bg).any():
        bad = torch.nonzero(n < min_bg, as_tuple=False)
        raise RuntimeError(
            f"Too few background pixels in frames: {[(int(v), int(t)) for v, t in bad.tolist()]}"
        )

    # Sufficient statistics
    sx = (X * M).sum(dim=-1)
    sy = (Y * M).sum(dim=-1)
    sxx = (X * X * M).sum(dim=-1)
    sxy = (X * Y * M).sum(dim=-1)

    # Closed-form least squares for a, b
    eps = 1e-8
    mx = sx / n
    my = sy / n
    varx = sxx / n - mx * mx
    cov = sxy / n - mx * my

    a = cov / (varx + eps)  # [V,T]
    b = my - a * mx

    # Apply alignment
    aligned_moge = (a[..., None] * X + b[..., None]).view(V, T, H, W)

    # Optionally save scale/shift
    scale = a.float().cpu()
    shift = b.float().cpu()

    # ---- Merge FG/BG ----
    final_depths = torch.where(bg_mask, D_bg_exp, aligned_moge)  # [V,T,H,W]

    # ---- Confidence map: high for BG, moderate for FG ----
    final_confs = torch.zeros_like(final_depths)
    final_confs[bg_mask] = 1000.0
    final_confs[~bg_mask] = 10.0

    # ---- Cache results ----
    np.save(final_depths_path, final_depths.cpu().numpy())
    np.save(final_confs_path, final_confs.cpu().numpy())
    np.save(os.path.join(cache_root, "scale.npy"), scale.cpu().numpy())
    np.save(os.path.join(cache_root, "shift.npy"), shift.cpu().numpy())

    return final_depths.unsqueeze(2).cpu(), final_confs.unsqueeze(2).cpu()


def _static_bg_mask_from_window(
        rgbs: torch.Tensor,
        win: int = -1,
        r: int = 7,  # spatial patch radius -> (2r+1)x(2r+1)
        diff_thresh: float = 10.0  # uint8 scale threshold
):
    """
    Fast BG detector using 3D max-pooling over frame-to-frame diffs.
    """
    V, T, C, H, W = rgbs.shape
    device = rgbs.device

    if T == 1:
        return torch.ones((V, T, H, W), dtype=torch.bool, device=device)

    if win == -1:
        win = T

    # 1) Frame-to-frame abs diff (channel-mean): boundaries of length T-1
    x = rgbs.float()
    diffs = (x[:, 1:] - x[:, :-1]).abs().mean(dim=2)  # [V, T-1, H, W]
    diffs = diffs.unsqueeze(1)  # [V, 1, T-1, H, W]  (N,C,D,H,W for 3D pool)

    # 2) 3D max pool over time & space:
    #    - temporal kernel spans (2*win-1) boundaries
    #    - spatial kernel spans (2r+1)x(2r+1) patch
    kt = max(1, 2 * win - 1)
    kh = kw = 2 * r + 1
    pt = (kt - 1) // 2
    ph = pw = r
    pooled = F.max_pool3d(diffs, kernel_size=(kt, kh, kw), stride=1, padding=(pt, ph, pw))
    pooled = pooled[:, 0]  # [V, T-1, H, W]

    # 3) Map boundary maxima back to frame centers (symmetric nearest-window approx)
    change = torch.zeros((V, T, H, W), device=device, dtype=pooled.dtype)
    change[:, 0] = pooled[:, 0]
    change[:, 1:-1] = torch.maximum(pooled[:, :-1], pooled[:, 1:])
    change[:, -1] = pooled[:, -1]

    # 4) Threshold -> background
    bg_mask = (change < diff_thresh)
    return bg_mask


def _moge_depths(seq_name, rgbs, cache_root, resize_to=512, batch_size=18):
    """Runs (and caches) MoGe-2; returns [V,T,H,W] float32 at native resolution."""

    # pip install git+https://github.com/microsoft/MoGe.git
    from moge.model.v2 import MoGeModel as MoGe2Model

    depths_path = os.path.join(cache_root, "moge_depths.npy")
    if os.path.isfile(depths_path):
        logging.info(f"Loading cached MoGe-2 depths for {seq_name} from {depths_path}")
        return torch.from_numpy(np.load(depths_path)).float()

    V, T, C, H, W = rgbs.shape
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = MoGe2Model.from_pretrained("Ruicheng/moge-2-vitl-normal").to(device).eval()

    if resize_to is None:
        h1, w1 = H, W
    else:
        if H >= W:
            h1, w1 = int(resize_to), max(1, round(resize_to * W / H))
        else:
            w1, h1 = int(resize_to), max(1, round(resize_to * H / W))

    imgs = rgbs.view(V * T, C, H, W).float()
    if (h1, w1) != (H, W):
        imgs = F.interpolate(imgs, size=(h1, w1), mode="bilinear", align_corners=False)
    imgs = (imgs / 255.0).to(device, non_blocking=True)  # [N,3,h1,w1]

    out_small = torch.empty((V * T, h1, w1), dtype=torch.float32, device=device)

    with torch.inference_mode(), torch.autocast(device_type=device.type, dtype=torch.bfloat16,
                                                enabled=(device.type == "cuda")):
        N = imgs.shape[0]
        for i in range(0, N, batch_size):
            chunk = imgs[i:i + batch_size]  # [b,3,h1,w1]
            pred = model.infer(chunk)  # expects batched input
            assert isinstance(pred, dict) and "depth" in pred, "MoGe-2 infer() must return dict with 'depth'."
            d = torch.as_tensor(pred["depth"], device=device)
            assert d.ndim == 3 and d.shape[0] == chunk.shape[0] and tuple(d.shape[1:]) == (h1, w1), \
                f"Depth shape {tuple(d.shape)} != ({chunk.shape[0]},{h1},{w1})"
            out_small[i:i + chunk.shape[0]] = d

    if (h1, w1) != (H, W):
        out = F.interpolate(out_small[:, None], size=(H, W), mode="bilinear", align_corners=False)[:, 0]
    else:
        out = out_small
    out = out.clamp_min(0).view(V, T, H, W).cpu()

    np.save(depths_path, out.numpy())
    return out


```

**Comment by Z1hanW** [comment]
I can see several failure in Dust3R, did you preset camera params (pose, intrinsics) during global optimization? That empirically helps, please see [here](https://github.com/naver/dust3r/issues/30)
